Revising and clarifying content (#10106)
keycloak/keycloak: docs/guides/src/main/server/configuration.adoc

title="Configuring Keycloak"
summary="An overview of the server configuration">

This guide describes the core concepts of Keycloak server configuration and server configuration options. This information includes configuration guidelines for optimizing server operation for faster startup and low memory footprint.

== Server Command Options

> You can view the server command options by adding the `help` option for individual commands:

Author: Okay, thanks for the clarification.  How about "Server options" with "Understanding."   I am suggesting standardizing on

- nouns for concepts and references sections 
- gerunds for task-oriented sections and guide names

* Database specific resources are configured and prepared to run against a certain database vendor
* By persisting build options into the server image, the server does not perform any additional step to interpret configuration options and (re)configure itself

.Run the `build` command to set the database to PostgreSQL before startup:
<@kc.build parameters="--db=postgres"/>

Once you run the `build` command, you do not need to set the same **build options** when you start the server.

=== Second stage: Starting Keycloak

The second stage involves starting the server by using the `start` command and including **configuration options** with the appropriate values.

.Run the `start` command to start the server while setting various database options
<@kc.start parameters="--db-url-host=keycloak-postgres --db-username=keycloak --db-password=change_me --hostname mykeycloak.acme.com"/>

Note that if you invoke commands containing special shell characters such as `;` using the CLI, you need to escape those characters. In that situation, you might choose to use the `keycloak.conf` file to set configuration options instead.

== Optimization by using a configuration file

> Most optimizations to startup and memory footprint can be achieved by using the `build` command. Additionally, you can use the `conf/keycloak.conf` file to set  configuration options. Using this file avoids some necessary parsing steps when providing configuration options using the CLI or environment variables.

Reviewer: ```suggestion
Most optimizations to startup and memory footprint can be achieved by using the `build` command. Additionally, you can use the `conf/keycloak.conf` file to set configuration options. Using this file avoids some necessary parsing steps when providing configuration options using the CLI or environment variables.
```

on environment variables to configure the server. However, you are not limited to a single configuration source or format.

== Configuration commands and stages

The server options are narrowed to a specific command or configuration stage. The goal is to perform a series of optimizations in a specific order to achieve optimal startup and runtime performance. This configuration occurs in two stages:

First stage:: Configuration performed before starting the server in order to build an optimized server image for use at runtime
Second stage:: Configuration performed as part of starting the server

[discrete]
=== The first stage

The first stage involves running the `build` command and setting any **build option** available from this command to configure the server.

The configuration options at <@links.server id="all-config"/> includes options that are marked with an icon. Those options are build options; they only take effect when you run the `build` command.
You can also check which options require a build by looking at the `help` message of the `build` command:

<@kc.build parameters="--help"/>

> The `build` command can produce an immutable and optimized server image, which is similar to building a container image. In addition to persisting build options, this command also performs optimizations for the best startup and runtime performance. The result is that much processing for starting and running the server is eliminated so that the server starts and runs faster.

Reviewer: it is not eliminated, just done beforehand in the build step.

> The `build` command can produce an immutable and optimized server image, which is similar to building a container image. In addition to persisting build options, this command also performs optimizations for the best startup and runtime performance. The result is that much processing for starting and running the server is eliminated so that the server starts and runs faster.

Author: Good. Thanks.

The following are some optimizations performed by the `build` command:

* Closed-world assumption about installed providers, which means that no need exists to re-create the registry at every startup
* Configuration files are pre-parsed to reduce I/O when starting the server
* Database specific resources are configured and prepared to run against a certain database vendor
* By persisting build options into the server image, the server does not perform any additional step to interpret configuration options and (re)configure itself

.Run the `build` command to set the database to PostgreSQL before startup:
<@kc.build parameters="--db=postgres"/>

Once you run the `build` command, you do not need to set the same **build options** when you start the server.

> [discrete]

Reviewer: see above, please avoid discrete

> === The second stage

Reviewer: ```suggestion
=== Second stage: Starting Keycloak
```

The second stage involves starting the server by using the `start` command and including **configuration options** with the appropriate values.

.Run the `start` command to start the server while setting various database options
<@kc.start parameters="--db-url-host=keycloak-postgres --db-username=keycloak --db-password=change_me --hostname mykeycloak.acme.com"/>

> Note that if you invoke commands containing special shell characters such as `;` using the CLI, you need to escape those characters.  In that situation, you might choose to use the `keycloak.conf` file to set configuration options instead.

Reviewer: ```suggestion
Note that if you invoke commands containing special shell characters such as `;` using the CLI, you need to escape those characters. In that situation, you might choose to use the `keycloak.conf` file to set configuration options instead.
```

== Optimization by using a configuration file

> In addition to using the `build` command instead of the `start` command for optimization, a second alternative exists. You can use the `conf/keycloak.conf` file to set environment variables or configuration options. For example, you might want to avoid using CLI options when running the `start` command.

Reviewer: ```suggestion
Most optimiziations to startup and memory footprint can be achieved by using the `build` command. Additionally, you can use the `conf/keycloak.conf` file to set  configuration options. Using this file avoids some necessary parsing steps when providing configuration options using the CLI or environment variables.
```

Reviewer: The edit did not make much sense because it changed the previously right contents. Tried  to come up with a better one.

Author: Okay, thanks, Dominik for the correction.

.Start the server
<@kc.start/>

By using the `keycloak.conf` file, the server can omit some steps at startup. As a result, the server starts faster.

> == Two configuration stages from one command

Reviewer: I'd like to leave the "auto-build" in the headline. This will be the searchterm everybody will search for who really wants to use this, as it is mentioned everywhere form CLI help to blog entries etc.

Suggestion: Leave as is, or something like "Using the `auto-build` option to automatically detect when the server needs a build" (which is too long imo)

Under certain circumstances, you might prefer to allow a longer startup time in favor of updating the values of build options when starting the server. Using the `auto-build` option, you can perform the two configuration stages by using a single command.

You start the server by entering the following command:

.Using the `auto-build` option
<@kc.start parameters="--auto-build --db postgres --db-url-host keycloak-postgres --db-username keycloak --db-password change_me --hostname mykeycloak.acme.com"/>

> By including the `auto-build` option, the server calculates the build options that have changed and runs the `build` command, if necessary, before starting the server.

Reviewer: ```suggestion
By including the `auto-build` option, the server calculates the build options that have changed and runs the `build` command, if necessary, before starting the server. Be aware that using `auto-build` likely more than doubles the time Keycloak needs to start up, so for most environments this approach is not optimal.
```

Reviewer: just to make that very clear

<@kc.start rootParameters="-cf myconfig.conf"/>

Changes to any *build option* defined in the `keycloak.conf` file that is targeted for the `build` command are ignored
if the value differs from the value for the last `build` command. In this case, make sure you run the `build` command again so that
any build option is updated accordingly.

=== Development versus production mode

The server supports the following operating modes:

Development mode:: This mode is activated every time you run the `start-dev` command. In this mode, some key configuration options are set to make it possible to start the
server for development purposes without the burden of having to define additional settings that are mandatory for production.

> Production mode:: This mode is activated by default when you run the `build` or the `start` command. Use this mode to set any configuration option that

Reviewer: ```suggestion
Production mode:: This mode is activated when you run the `build` or the `start` command. Use this mode to set any configuration option that
```

Reviewer: as we don't have profile support anymore, the "by default" can go away, as this is _the_ mode when invoking start / build now.

is needed for deploying Keycloak in production.

By default, the configuration options for the production mode are commented out in the `conf/keycloak.conf`. These examples
 are meant to give you an idea about the main settings to consider when running in production.

== Unsupported server options

In most cases, the available options from the server configuration should suffice to configure the server.
However, you might need to use properties directly from Quarkus to enable a specific behavior or capability that is missing from the server configuration.

As much as possible, avoid using properties directly from Quarkus. If your need is essential, consider opening an https://github.com/keycloak/keycloak/issues/new?assignees=&labels=kind%2Fenhancement%2Cstatus%2Ftriage&template=enhancement.yml[issue] first and help us
to improve the server configuration.

To configure the server using Quarkus properties, perform the following steps:

> . Create a `conf/quarkus.properties` file and define any property you need.

Reviewer: Not 100% sure, but shouldn't these stay bullet points of an unordered list? (a `*`)

Reviewer: you are right, was a bit confused, sorry. so when you are changing the whitespace, feel free to also change back to step / dots.
